File: boj/b_11054.js
function solution(n, array) {
  const [tableA, tableB] = [Array(n).fill(1), Array(n).fill(1)];
  for (let i = 0; i < n; i += 1) {
    for (let j = 0; j < i; j += 1) {
      if (array[i] > array[j]) tableA[i] = Math.max(tableA[i], tableA[j] + 1);
    }
  }
  for (let i = n - 1; i >= 0; i -= 1) {
    for (let j = n - 1; j > i; j -= 1) {
      if (array[i] > array[j]) tableB[i] = Math.max(tableB[i], tableB[j] + 1);
    }
  }
  let result = 0;
  for (let i = 0; i < n; i += 1) {
    result = Math.max(result, tableA[i] + tableB[i] - 1);
  }
  return result;
}
console.log(solution(10, [1, 5, 2, 1, 4, 3, 4, 5, 2, 1]));

File: boj/b_11054.py
import sys
sys.stdin = open("input.txt", "r") # 제거
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
table_a, table_b = [1] * n, [1] * n
for i in range(n):
    for j in range(i):
        if arr[i] > arr[j]:
            table_a[i] = max(table_a[i], table_a[j] + 1)
for i in range(n - 1, -1, -1):
    for j in range(n - 1, i, -1):
        if arr[i] > arr[j]:
            table_b[i] = max(table_b[i], table_b[j] + 1)
result = 0
for i in range(n):
    result = max(result, table_a[i] + table_b[i] - 1)
print(result)
